### ISSUE TITLE
bug: Memory issue in whisper plugin

### ISSUE THREAD

**Issue by apepkuss** [WasmEdge/WasmEdge#3912]
### Summary

A potential memory issue is found while running the [test code](https://github.com/second-state/WasmEdge-WASINN-examples/blob/test-whisper/whisper-basic/src/main.rs): In the period of running 7 wav files (6/7 in 19.2MB, 1/7 in 2.9MB), observed that the memory consumptions increase each round, that is, it seems that the memory used each round is not released after each round is done. Approx. 4% or more memory consumption is increased after each round. The table below shows the statistics of memory consumptions of each round. Note that before starting the test code, the memory consumption is 38% of the total memory of the test machine (Apple M1 Pro + 32G Mem + macOS-15.1.1).

![image](https://github.com/user-attachments/assets/eaabace0-3331-48c8-a3a5-ebe1af6e5d22)

- The two rows in red color reach the 83% of total memory. It could be the maximum memory consumption allowed by macOS.
- The row in yellow color represents that all wav files have been processed; and sleeping 40 seconds for observing the memory consumption.
- The row in green color shows the memory consumption after the test code exits.

From the last column, we can see that the memory consumption increases after each round. After 7 rounds, the memory consumption climbed up to 79% ~ 82%.

To illustrate the complete test, a YouTube video is shared [here](https://youtu.be/UP0asxlXEc8). To reproduce the issue, the test code is [here](https://github.com/second-state/WasmEdge-WASINN-examples/blob/test-whisper/whisper-basic/src/main.rs). The command to run the test code is

```bash
WASMEDGE_PLUGIN_PATH=/Users/sam/wasmedge-plugin-registry/whisper/d259eed/plugin wasmedge --dir .:. ./target/wasm32-wasip1/release/whisper-basic.wasm ggml-medium.bin
```
 **Note** that the wav files used by the test code can be found in  the **Screenshots** section below.

### Reproduction steps

N/A

### Screenshots

The wav files used by test code:
[output_000.wav.zip](https://github.com/user-attachments/files/18093239/output_000.wav.zip)
[output_001.wav.zip](https://github.com/user-attachments/files/18093242/output_001.wav.zip)
[output_002.wav.zip](https://github.com/user-attachments/files/18093244/output_002.wav.zip)
[output_003.wav.zip](https://github.com/user-attachments/files/18093273/output_003.wav.zip)
[output_004.wav.zip](https://github.com/user-attachments/files/18093249/output_004.wav.zip)
[output_005.wav.zip](https://github.com/user-attachments/files/18093252/output_005.wav.zip)
[output_006.wav.zip](https://github.com/user-attachments/files/18093258/output_006.wav.zip)

### Components

Others

### WasmEdge Version or Commit you used

0.14.1

### Operating system information

macOS 15.1.1

### Hardware Architecture

Apple Silicon M1 Pro

**Comment by PeterD1524** [comment]
Maybe call `graph.unload` at the end of each iteration or reuse the ctx?

`Graph` should probably implement `Drop` and call the system call unload inside it. However, not every wasi-nn plugin implements unload.

Other problems:
1. The Rust API allows using `GraphExecutionContext` after the corresponding `Graph` being unloaded which will lead to SIGSEGV. Maybe `GraphExecutionContext` should hold a reference to `Graph`.
2. There is no API to unload a `GraphExecutionContext`. If I understand correctly they are not released in `graph.unload`.

**Comment by apepkuss** [comment]
> Maybe call `graph.unload` at the end of each iteration or reuse the ctx?

The motivation of the test code mentioned above is to avoid reusing `ctx`. It's just intended to verify if a new graph (also ctx) can work correctly for each round.

> 1. The Rust API allows using `GraphExecutionContext` after the corresponding `Graph` being unloaded which will lead to SIGSEGV. Maybe `GraphExecutionContext` should hold a reference to `Graph` so it cannot be used after `Graph` dropped.

If `GraphExecutionContext` holds a reference to a `Graph` instance, like the code below, it would become difficult to declare a `static` global variable to hold the `Graph` instance. If I am wrong, please correct me.

```rust
pub struct GraphExecutionContext<'a> {
    graph: &'a Graph,
    ctx_handle: syscall::GraphExecutionContextHandle,
}
```

> 2. There is no API to unload a `GraphExecutionContext`. If I understand correctly they are not released in `graph.unload`.

Yeah, you're right. I updated the test code with two lines: [line.74 - 75](https://github.com/second-state/WasmEdge-WASINN-examples/blob/test-whisper/whisper-basic/src/main.rs#L73-L75). These two lines explicitly drop the `ctx` and `graph` in the order. According to another test run, the `drop` actions are conducted, but do not do what they should do: releasing the memory. As you pointed out, the `drop` does not trigger the low-level `unload` function. So, a new question appears: where is the `unload` function? for both `Graph` and `GraphExecutionContext`.

**Comment by apepkuss** [comment]
Great! I'll try it, thanks a lot!

**Comment by PeterD1524** [comment]
> If `GraphExecutionContext` holds a reference to a `Graph` instance, like the code below, it would become difficult to declare a `static` global variable to hold the `Graph` instance. If I am wrong, please correct me.

Yes. You also would not be able to store a Graph and its GraphExecutionContext in a same struct like in https://github.com/LlamaEdge/LlamaEdge/blob/a1d000dd8a9d06c95a07985f0eb1e8df6432dd1d/crates/llama-core/src/graph.rs#L217-L222 because it's [self-referential](https://stackoverflow.com/a/32300133).

I guess these are some trade-offs that wasmedge-wasi-nn should consider. Maybe using Rc or Arc to Graph instead of reference will avoid these issues. Not sure what's the best approach.

**Comment by apepkuss** [comment]
> Yes. You also would not be able to store a Graph and its GraphExecutionContext in a same struct like in https://github.com/LlamaEdge/LlamaEdge/blob/a1d000dd8a9d06c95a07985f0eb1e8df6432dd1d/crates/llama-core/src/graph.rs#L217-L222 because it's [self-referential](https://stackoverflow.com/a/32300133).
> I guess these are some trade-offs that wasmedge-wasi-nn should consider. Maybe using Rc or Arc to Graph instead of reference will avoid these issues. Not sure what's the best approach.

`llama-core` depends on the `ggml` branch of `wasmedge-wasi-nn`, in which  [GraphExecutionContext](https://github.com/second-state/wasmedge-wasi-nn/blob/ggml/rust/src/graph.rs#L252-L254) removes the reference to `Graph` instance. We maintain the dependency in a higher-lever `Graph` object defined in `llama-core`. As you said, this is a `trade-off` design for our use case.

> There is no API to unload a GraphExecutionContext. If I understand correctly they are not released in graph.unload.

I tried `graph.unload`. According to the test, the memory issue is still there. My guess is that `graph.unload` is only responsible for releasing the memory used by `graph`, and it seems still necessary to have `ctx.unload` to release the memory used by `ctx` (at least in the case of whisper plugin).

**Comment by PeterD1524** [comment]
I opened a [PR](https://www.github.com/WasmEdge/WasmEdge/pull/3917) for adding a new wasi-nn function `finalize_execution_context` to unload ctx. If this PR is merged, maybe we can let the whisper plugin implement it.

**Comment by hydai** [comment]
I think it could be an implementation problem inside the Whisper plugin. We should allocate memory inside the `Graph` and release this memory during `graph.unload`.

Adding a new method for finalizing the execution context is fine with me, but I just want to make sure if this is an implementation issue or if it's a design issue that requires us to introduce this new method.

**Comment by apepkuss** [comment]
In my view, the execution context needs its own `unload`. According to the implementation of [graph.unload](https://github.com/second-state/wasmedge-wasi-nn/blob/ggml/rust/src/graph.rs#L235), it calls a low-level function `syscall::unload`, which only works with a graph handler. However, there is no low-level function to unload [ctx_handle](https://github.com/second-state/wasmedge-wasi-nn/blob/ggml/rust/src/graph.rs#L253C5-L253C15). In addition, `GraphExecutionContext` has a method named [fini_single](https://github.com/second-state/wasmedge-wasi-nn/blob/ggml/rust/src/graph.rs#L290-L292), which calls `syscall::fini_single` to release `ctx_handle`. As I know, the `fini_single` method is only for clearing the context in the case of chat completion tasks in streaming mode. Maybe this method could also be seen as a "proof" that we need a separate method to release an execution context instance as the `fini_single` does.

**Comment by PeterD1524** [comment]
I think it is possible to create multiple execution contexts for the same graph (I don't know how will that be useful), so maybe users wants to keep the graph and free some of its execution context?

I think it's okay to release the memory for execution context in `graph.unload` because execution contexts cannot live longer than their graph. However, if a new method `context.unload` is added, it should be clear to users to not call `context.unload` after calling `graph.unload` otherwise they will unload with a stall context id.

Some information I think may be useful:

I checked 3 implementations of graph.unload:
1. ggml
    - calls `llama_free_model` for something stored on graph
    - erases the graph struct from `Env.NNGraph`
    - does not erase its context structs from `Env.NNContext`
3. whisper
    - calls `whisper_free` for something stored on graph
    - erases the graph struct from `Env.NNGraph`
    - does not erase its context structs from `Env.NNContext`
4. chattts
    - wrongly using graph id parameter as context id (and use it to get context and then get its graph)
    - calls `Py_XDECREF` for something stored on graph
    - does not erase the graph struct from `Env.NNGraph`
    - does not erase its context structs from `Env.NNContext`

I actually do not fully understand what the role is for execution context. There was even an attempt to eliminate GraphExecutionContext (https://www.github.com/WebAssembly/wasi-nn/issues/43).

**Comment by apepkuss** [comment]
@hydai @PeterD1524 I would like to make sure that what we're discussing on the issue itself and potential solutions is based on `ggml` branch of [wasmedge-wasi-nn](https://github.com/second-state/wasmedge-wasi-nn) project，right?

**Comment by hydai** [comment]
Correct. However, we should prepare the necessary information and report to the upstream to discuss a more general specification.

**Comment by ANAMASGARD** [comment]
Hi @apepkuss @hydai @PeterD1524 — quick update on #3912.

We’ve opened PR #5057 (linked above) with the WasmEdge side of the fix:
 - Whisper unload now cleans up remaining execution contexts before freeing the model
 - finalize_execution_context tests added for the Whisper backend (mirroring BitNet)
 - Local verification: without finalize + unload, RSS climbed ~2.3GB over 7 rounds; with the fix, growth was ~40MB and stable

The PR is in draft for now due to the one-PR limit on my side. When you have a moment, I would appreciate a review — and if it looks good, converting it from draft to ready for review would be great, Thank You so much .
